Added support for manga list and anime/manga list entry update

naps250 opened the pull request:
Added support for user's manga list.
Added support for updating user's anime and manga MAL records.
Minor renaming and reordering.
Added 'Anime' and 'Manga' folders (and namespaces) for better organization of the 'MalApi' project.

naps250 committed:
Added support for user's manga list.
Added support for updating user's anime and manga MAL records.
Minor renaming and reordering.
Added 'Anime' and 'Manga' for better organization of the 'MalApi' project.

LHCGreg commented:
Thanks for your contribution. I've glanced through the changes. I'll take a deeper look this weekend and will have more comments then but I have some comments now.

First, the creation of the MalApi.Anime and MalApi.Manga namespaces. I can see how one might like how that organizes the MalApi code, especially with several parallel anime and manga types, but I don't know if that's something a library client would want. Now you pretty much have to import two namespaces to do anything. There is not such a huge number of types in this library and moving some types into another namespace is a backwards compatibility breaking change. Some breaking changes I think will be necessary like the renaming of CompletionStatus -> AnimeCompletionStatus due to my lack of foresight, but it should be kept to a minimum.

Secondly, tests! The new functionality should have some coverage in unit/integration tests. I realize integration tests for functionality that makes updates to a user's list poses some problems regarding credentials, but it can be done, for example by reading the username and password from environment variables and documenting what kind of setup is required to run the tests as far as setting the environment variables and setting up the MAL account used in the tests.

naps250 committed:
Reverted the namespace change but still kept the folders.
Removed unused using statements.

naps250 commented:
Hi,

Thank you for you feedback. I really appreciate it.

I reverted the namespace change but still kept the new folders. I had completely overlooked the backward compatibility point of view when initially making this change.
I also went through the .cs files and removed any unused using statements.

Both of the above are part of the second commit.

I will start working on the test now while waiting for your further comments.

LHCGreg commented:
Please revert the removal of unused using directives. The changes clutter the PR and are unrelated to manga and update functionality. I find it useful to have common namespaces like System.Collections.Generic and System.Linq already imported.

LHCGreg commented on MalApi/Anime/UpdateAnimeObject.cs:
```diff
@@ -0,0 +1,85 @@
+namespace MalApi
+{
+    [XmlRoot("entry")]
+    public class UpdateAnimeObject : UpdateObjectBase
```
"Object" is kind of generic. "AnimeUpdate" would describe the purpose as well.

LHCGreg replied:
The properties in this class are all strings. The users of the library should be dealing with integers for episode count, AnimeCompletionStatus for status, etc, instead of expecting them to serialize those properties to strings.

LHCGreg replied:
What does it mean if a property is not set? Does it not get updated or does it get set to an empty value in MAL? This should be documented in a doc comment on the class.

LHCGreg replied:
Might be nice to link to the MAL documentation at https://myanimelist.net/modules.php?go=api#animevalues in a comment.

naps250 replied:
I agree with all your comments.

LHCGreg commented on MalApi/Anime/UpdateAnimeObject.cs:
```diff
@@ -0,0 +1,85 @@
+                                 string enableRewatching = null,
+                                 string comments = null,
+                                 string tags = null)
+        {
```
Is this constructor adding any value over calling the default constructor and then setting properties?

naps250 replied:
Not really... I'll remove it.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -227,6 +305,147 @@ public Task<MalUserLookupResults> GetAnimeListForUserAsync(string user)
+        /// <param name="user">MAL user</param>
+        /// <param name="password">MAL password</param>
+        /// <returns></returns>
+        public async Task<string> UpdateAnimeForUserAsync(int animeId, UpdateAnimeObject updateInfo, string user, string password, CancellationToken cancellationToken)
```
The return type should be `Task` instead of `Task<string>` with an always null string.

LHCGreg replied:
Please change "user" to "username" to be more clear.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -16,7 +15,14 @@ namespace MalApi
+        private const string MalAppAnimeInfoUri = "https://myanimelist.net/malappinfo.php?status=all&type=anime&u={0}";
+        private const string MalAppMangaInfoUri = "https://myanimelist.net/malappinfo.php?status=all&type=manga&u={0}";
+
+        private const string MalAppAnimeUpdateUri = "https://myanimelist.net/api/animelist/update/{0}.xml";
```
No need for "App" in the name since it has nothing to do with malappinfo.php. Also, this is a format string rather than a string that can be used directly, but the name does not indicate this. Consider moving the format string directly into the function that uses it so that there is no distance between the format string and the string.Format call. That way it is clear where the format arguments get put and how many are needed.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -227,6 +305,147 @@ public Task<MalUserLookupResults> GetAnimeListForUserAsync(string user)
+        /// <returns></returns>
+        public async Task<string> UpdateAnimeForUserAsync(int animeId, UpdateAnimeObject updateInfo, string user, string password, CancellationToken cancellationToken)
+        {
+            string userInfoUri = string.Format(MalAppAnimeUpdateUri, Uri.EscapeDataString(animeId.ToString()));
```
Uri.EscapeDataString is not really necessary, but ok.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -227,6 +305,147 @@ public Task<MalUserLookupResults> GetAnimeListForUserAsync(string user)
+                    catch (Exception ex)
+                    {
+                        throw new Exception(string.Format("Exception for user {0} and anime ID {1}.", user, animeId), ex);
+                    }
```
responseProcessingFunc should not be catching exceptions and rethrowing them as a general Exception. The baseErrorMessage parameter to ProcessRequestAsync is used as part of the error handling in ProcessRequestAsync.

LHCGreg replied:
https://myanimelist.net/modules.php?go=api says the response is 'Updated' or detailed error message. Have you tested this? If there is an error does it return an HTTP error status code or does it return an OK HTTP status code? If there is an error message, is it a plain string or is it in something like xml or json? It looks like the code here does nothing to surface the MAL error message to the caller.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -227,6 +305,147 @@ public Task<MalUserLookupResults> GetAnimeListForUserAsync(string user)
+                {
+                    try
+                    {
+                        return xmlTextReader.ToString();
```
If responseProcessingFunc has nothing useful to return, which makes sense for an update API call, responseProcessingFunc should have a return type of object and return null.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
             return request;
         }
 
+        private HttpRequestMessage InitNewRequestWithCredentials(string user, string password, string uri, HttpMethod method, UpdateObjectBase updateInfo = null)
```
Consider moving uri and method to be the first two parameters, as they are kind of the "primary" parameters of the method.

LHCGreg replied:
updateInfo does not belong in InitNewRequestWithCredentials. When functionality for adding and removing of anime/manga entries is added, that will use credentials but will not take the same input that updating does. Any MAL endpoint-specific logic can be done in the method for that endpoint on top of the HttpRequestMessage object returned by InitNewRequest or InitNewRequestWithCredentials.

naps250 replied:
Technically, delete takes only an id as input while add and update take the same input as far as I understand from the API. But again I agree that setting the content inside InitNewRequestWithCredentials is not the best solution due to possible future changes to the API and the functionality that the method name reflects.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
+
+            if (UserAgent != null)
+            {
+                request.Headers.TryAddWithoutValidation("User-Agent", UserAgent);
```
You can avoid repeating the logic that is in InitNewRequest by calling InitNewRequest and then adding onto the request it returns.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
+            {
+                request.Headers.TryAddWithoutValidation("User-Agent", UserAgent);
+                // Adding the authorization header with the credentials
+                request.Headers.TryAddWithoutValidation("Authorization", "Basic " + credentials);
```
Why only set the Authorization header if the UserAgent property is set?

LHCGreg replied:
Using header properties that the HttpRequestMessage API supports directly is preferable to using TryAddWithoutValidation. TryAddWithoutValidation is used for User-Agent because it complains if the user agent you set does not follow the same format browsers use, which is not the case when it is being (mis)appropriated as an API key. For authorization you should be able to use

```
request.Headers.Authorization = new AuthenticationHeaderValue("Basic", credentials);
```

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
+            if (user != null && password != null)
+            {
+                var plainTextBytes = Encoding.UTF8.GetBytes(user + ":" + password);
+                credentials =  Convert.ToBase64String(plainTextBytes);
```
There's an extra space after the equals sign.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
+            }
+            else
+            {
+                throw new Exception("Invalid input credentials format (both username and password have to be input OR just the base64 representation");
```
Instead of Exception, ArgumentNullException would be more appropriate. In order to call the ArgumentNullException constructor that takes a parameter name, you'll have to check user and password one at a time.

LHCGreg replied:
Also, I am confused by the "OR just the base64 representation" in the error message.

naps250 replied:
The error message part was a leftover from earlier ideas.

LHCGreg commented on MalApi/MyAnimeListApi.cs:
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
+			{
+				var xml = updateInfo.GenerateXml();
+				var content = "data=" + WebUtility.UrlEncode(xml);
+				request.Content = new StringContent(content, Encoding.UTF8, "application/x-www-form-urlencoded");
```
Does this not work? The FormUrlEncodedContent class exists so you don't have to encode it yourself.

```
request.Content = new FormUrlEncodedContent(new KeyValuePair<string, string>[] { new KeyValuePair<string, string>("data", xml) });
```

LHCGreg commented:
Added comments for anime update functionality. Most or all of it applies to manga update functionality as well.

naps250 committed:
Using directives removal revert.

naps250 committed:
Addressed almost all of your comments.

Regarding the response from the update: there are two ways to break the XML - in the first scenario it returns a message for invalid XML and in the other it a MySQL error and status OK ("Updated"). Unfortunately the response does not contain the MySQL information bit. On the other hand I don't know how you can break the XML with the current configuration.

When credentials are not correct the already existing code takes care of it and returns a 401.

naps250 committed:
Added some tests.

Added integration tests for the update functionality. I apologize for the poor execution but my experience with tests is very limtied.

There is some strange occurance where some of the tests fail depending on the name of the Initialize function. I guess the name somehow is responsible for the order of execution but I could not find information about it.

naps250 commented:
Thanks again for your feedback.

I addressed all of your comments regarding the functionality itself but couldn't quite get the tests set up properly. My apologies!

LHCGreg commented on MalApi/MalAppInfoXml.cs:
```diff
@@ -194,6 +202,100 @@ public static MalUserLookupResults Parse(XDocument doc)
+
+            XElement myinfo = GetExpectedElement(doc.Root, "myinfo");
+            int userId = GetElementValueInt(myinfo, "user_id");
+            string canonicalUserName = GetElementValueString(myinfo, "user_name");
```
This first section of the method is duplicated from ParseAnimeResults. A way to avoid the duplication would be to add another constructor to MalUserLookupResults that takes both an anime list and a manga list and have one function in MalAppInfoXml that looks for both "anime" XML elements and "manga" XML elements.

MalUserLookupResults should then have an empty list for anime or manga instead of null.

LHCGreg commented on MalApi/Manga/MalMangaType.cs:
```diff
@@ -0,0 +1,14 @@
+        Manga = 1,
+        Novel = 2,
+        OneShot = 3,
+        Doujin = 4, // not sure
```
Why aren't you sure about Doujin and OEL?

smovsesyan-rms replied:
Because I couldn't find such entries when I added that. I found a doujin and confirmed that it returns 4, leaving OEL wtih 7, so this is confirmed OK now.

naps250 replied:
I didn't bother finding a MAL entry to verify it at the time of writing. I went back and found a doujin and confirmed that it returns 4. That leaves OEL as for 7, so this is OK.

LHCGreg commented on MalApi/Anime/UpdateAnime.cs:
```diff
@@ -0,0 +1,215 @@
+    /// More details: https://myanimelist.net/modules.php?go=api#animevalues
+    /// </summary>
+    [XmlRoot("entry")]
+    public class UpdateAnime : UpdateObjectBase, IXmlSerializable
```
Class names should be nouns, so this and UpdateManga should be AnimeUpdate/MangaUpdate.

LHCGreg replied:
I thought I had commented on this yesterday, I guess I missed it. Update functionality should use System.Xml.Linq for XML generation instead of System.Xml.Serialization:

- Consistency. The MalAppInfoXml class already uses System.Xml.Linq.
- It avoids adding a dependency on System.Xml.XmlSerializer to the nuget package.
- It avoids putting implementation details in the API surface of MalApi such the XML-related attributes and the UpdateObjectBase class.

The method for creating XML should be internal, not public, and would look something like this using System.Xml.Linq.

```
            XDocument document = new XDocument(
                new XDeclaration("1.0", "UTF-8", null),
                new XElement("entry",
                    new XElement("Episode", Episode),
                    new XElement("Status", (int?)Status),
                    new XElement("Score", Score),
                   ...
                )
            );

            using (Utf8StringWriter writer = new Utf8StringWriter())
            {
                document.Save(writer);
                return writer.ToString();
            }
```

LHCGreg commented:
I got a chance to look at the tests. Perhaps I was not clear when I mentioned environment variables. The idea is that username and password get stored in environment variables, *outside of the test code* so there is no risk of committing MAL credentials to git. If someone wants to run the tests, they are responsible for setting the environment variables (for example, using the [Windows environment variables editor](https://superuser.com/a/985947) outside of Visual Studio or [a .runsettings file](https://docs.microsoft.com/en-us/visualstudio/test/configure-unit-tests-by-using-a-dot-runsettings-file)). There would be an environment variable for username and password of a test account and the readme.txt would document what environment variables are required and what is expected of that test account in order for the tests to pass. For example, must have at least one anime and manga rating, must have an entry for Cowboy Bebop, whatever. So there do not need to be variables for an anime id, episode number, test score, or test status. Those are hardcoded in the tests and documented in the readme. Ideally the state of the test account when tests are not being run does not have a score on any anime or manga to avoid having any influence, no matter how small, on the overall average score.

It's not necessary to use the environment variable in the test for getting manga list if the test is just checking that the list returned is not empty. You can hardcode any user that has a non-empty manga list, like I did for the anime test. Alternatively the test could continue to use the environment variable test user and do some more thorough checks than if the list is non-empty.

The integration tests for updating should verify that the update actually happened by getting the user's list and checking that it is as expected.

The integration tests should pass no matter how many times they are run. That is, the act of running the tests should not put the test user's list into a state that causes the tests to fail the next time they are run.

The integration tests should also be meaningful tests no matter how many times they are run. If I run the update test and it only does one update, then if I run the test again, it's not really testing anything because the state of the user's list is still the same. A way to ensure that the update functionality is being tested is to do two updates and get the user's list after each.

The tests should cover as many scenarios as possible. If there is a bug with updating any particular property, they should catch that. They should catch a bug that happens when only some properties are specified in the update. They should cover not just the happy path but test that errors are handled correctly. For example, what happens if you update with a wrong password? Does the code correctly throw an exception of the correct type that indicates what was wrong? Have to be careful with this though, as MAL will IP block you for a bit if you give incorrect credentials too many times in a short period. How about if you give an invalid value for one of the properties?

There should be unit tests for the MalAppInfoXml parsing of manga lists, even if the integration tests are modified to do more than test that a returned list is non-empty, so that functionality can be tested without needing to set up a test MAL user or in case the integration tests are not able to be run for some reason. At the moment there is no test coverage for manga list XML parsing.

naps250 committed:
Replaced System.Xml.Serialization with System.Xml.Linq for XML generation.
Confirmed Doujin and OEL enum entries.
Removed UpdateObjectBase.
Put Utf8StringWriter in a separate class.
Renamed UpdateAnime and UpdateManga.

naps250 commented:
I addressed your latest comments and will focus on the tests now.

naps250 committed:
Missed to remove the comments.

naps250 commented:
In order to verify the updated information I will need to harvest some of it from the page's HTML since the API does not return everything. Do you prefer I continue to use regex like you've done for the genres or should I include a 3rd party library like the [Html Agility Pack](http://html-agility-pack.net/?z=codeplex)?

LHCGreg commented:
HtmlAgilityPack is fine. Since this is just used for testing (I assume), adding a dependency is no big deal. I try to avoid adding dependencies to MalApi itself when possible but if this code later becomes part of the MalApi's public API I think I'd be ok with HtmlAgilityPack as a dependency since it's relatively small and fits the purpose of MalApi well.

naps250 commented:
Your assumption is correct. This is needed purely for testing. I will go ahead and use HAP.

naps250 committed:
Added integration test for anime and manga updates (all the values and individual test for each property).
Added wrong username tests.
Added HtmlAgilityPack for use in tests.
Added HTML scraping tests for anime and manga.
Various bug fixes.
Changed the readme to include set up information for environment variables.
Added/renamed embedded XML files used for unit tests.
Commented out ScanGroup/scan_group property for manga since there doesn't seem to be a way to confirm that it's actually doing anything.

LHCGreg commented:
I thought the getting of anime/manga details for a user was going to be for testing only? Not only is it part of the MyAnimeListApi class, but it is public. Let's keep the changes to the public API in this pull request to updating only. Functionality for getting details for a particular anime/manga for a user has its own challenges:

* The MyAnimeListApi class is documented as having thread-safe methods, but logging in using cookies on an HttpClient that is a class member variable makes that no longer true.
* Logging in every time when getting anime/manga's details for a user is suboptimal.

In order to avoid adding a dependency on HtmlAgilityPack for functionality that won't be used by library users, the code for getting anime/manga's details for a user should be pulled out of the MyAnimeListApi class and into the integration test project. That code can be a little more straightforward if it doesn't need to use MyAnimeListApi's internal abstractions that do logging and exception handling.


Some other comments:

You can use

`Assert.Throws<MalApiRequestException>(() => api.UpdateAnimeForUser(animeId, partialBaseInfo, username, password));`

For asserting that some code throws a particular exception.

What happens if a user tries to update an anime id that doesn't exist or that isn't on their list? I didn't see any test coverage for that.

naps250 committed:
Updated XUnit packages because current version was no longer working (https://github.com/Microsoft/vstest/issues/811).
Removed HtmlAgilityPack (HAP) from the MalApi project.
Added HAP to the integration and unit test projects.
Moved update test logic to the integration test project.
Doubled the scraping methods in the unit test project to keep the scraping method tests and updated the readme.txt in integration test project to include a warning/notification.
Added test for non-existant MAL entry.
Updating id that is not on your list does nothing but still returns "Updated" response.
